Added event prop 'screen' to be recorded

octaedro opened the pull request:
Fixes #4633

The event prop `screen` wasn't being sent with the event `wcadmin_wcadmin_inbox_action_click`.
This PR adds the event prop `screen` to `wcadmin_wcadmin_inbox_action_click`.
This change will apply for all the call to action buttons of the admin notes.


### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Detailed test instructions:
- Open the Inbox Panel, choose a message, and press its call to action button.
- Verify that the screen name has been sent as a prop in the event `wc_admin_record_tracks_event` when a note's action is triggered. ([Here](https://github.com/woocommerce/woocommerce-admin/blob/16770825fe26015e32f52c75a8331502638af880/src/API/NoteActions.php#L133) is where we send it).
- Go to another page (E.g. the Analytics Overview, URL ->`/wp-admin/admin.php?page=wc-admin&path=%2Fanalytics%2Foverview`).
- Verify the sent screen name corresponds with the page name.
- Also, verify the screen name is sent with other types of admin notes.
E.g.:
![screenshot-one wordpress test-2020 06 19-13_17_53](https://user-images.githubusercontent.com/1314156/85155797-97e81d80-b22f-11ea-8e3e-9cdb9d32417c.png)

- If you don't have other types of admin notes available to test, you can add one executing these two sentences in the DB:
````
INSERT INTO `wp_wc_admin_notes` (`note_id`, `name`, `type`, `locale`, `title`, `content`, `content_data`, `status`, `source`, `date_created`, `date_reminder`, `is_snoozable`, `layout`, `image`, `is_deleted`, `icon`) VALUES (9999, 'wc-admin-test-note', 'update', 'en_US', 'Lorem ipsum note', 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed eiusmod tempor incidunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.', '{}', 'unactioned', 'woocommerce-admin', '2020-06-17 19:22:48', NULL, 0, 'plain', NULL, 0, 'info');
````
and
````
INSERT INTO `wp_wc_admin_note_actions` (`note_id`, `name`, `label`, `query`, `status`, `is_primary`, `actioned_text`) VALUES (9999, 'learn-more', 'Learn more', '?page=wc-addons&section=helper', 'unactioned', 0, '');
````

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:
Fix: Added event prop `screen` to `wcadmin_wcadmin_inbox_action_click`.
<!--- Optional: Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance:. If no note is entered, one will be constructed from the title and labels. --->

octaedro committed:
Added event prop to be recorded

This commit adds the event prop `screen` to `wcadmin_wcadmin_inbox_action_click`

joshuatf commented on src/API/NoteActions.php:
```diff
@@ -115,6 +119,8 @@ public function trigger_note_action( $request ) {
 			$tracks_event = 'wcadmin_inbox_action_click';
 		}
 
+		$screen_name = $request->get_param( 'screen' );
```
Could we not grab the referer from the request in here instead of manually sending it from the client?

octaedro replied:
Hi @joshuatf, thanks for the review.
I chose to send the param to try to have only one method where the screen names come from, but the solution you mentioned (take the referer from the request) is better. I applied it in the commits 8af0e02 and 3ff440f
In this PR I also moved the method `getScreenName` to a more general `utils` file. Now it is not directly affecting this PR but I'll leave this method there because I think that it's a better place for it.

joshuatf replied:
>In this PR I also moved the method getScreenName to a more general utils file. Now it is not directly affecting this PR but I'll leave this method there because I think that it's a better place for it.

Nice clean up!  Agreed, good change here.

joshuatf commented on src/API/NoteActions.php:
```diff
@@ -27,7 +27,7 @@ class NoteActions extends Notes {
 		register_rest_route(
 			$this->namespace,
-			'/' . $this->rest_base . '/(?P<note_id>[\d-]+)/action/(?P<action_id>[\d-]+)',
+			'/' . $this->rest_base . '/(?P<note_id>[\d-]+)/action/(?P<action_id>[\d-]+)/screen/(?P<screen>[a-z_-]+)',
```
If for some reason we can't use the referer, I think it might be better to include these params in the body of the request as opposed to the URL.

joshuatf reviewed (commented):
Very creative solution for tackling this, Fernando!  Left a couple comments that may possibly make this easier to manage, but let me know what you think.

octaedro committed:
Getting screen name from referer

Now the we get the screen name from the referer

octaedro committed:
The screen name is not being sent anymore to the API

The screen name is not sent anymore to the API. Also, the $screen_name is set as empty in the API to make the method more defensive.

joshuatf commented on src/API/NoteActions.php:
```diff
@@ -115,6 +115,24 @@ public function trigger_note_action( $request ) {
 			$tracks_event = 'wcadmin_inbox_action_click';
 		}
 
+		$screen_name = '';
```
Optional, but we can extract this block to its own function which might make it easier to follow or reuse going forward.  E.g., `get_screen_name()`.

octaedro replied:
Good idea! Sounds good to me, I modified this in the commit e5fdd11

joshuatf commented on src/API/NoteActions.php:
```diff
@@ -115,6 +115,24 @@ public function trigger_note_action( $request ) {
+		if ( isset( $queries ) ) {
+			$page      = isset( $queries['page'] ) ? $queries['page'] : null;
+			$path      = isset( $queries['path'] ) ? $queries['path'] : null;
+			$post_type = isset( $queries['post_type'] ) ? $queries['post_type'] : null;
```
This works well for post type list page, but we won't be able to see the type from individual posts this way.  For example, clicking into an individual order or product and actioning a note, the `screen` property will be blank.

I'm not sure how important this is for tracking (feel free to weigh in @pmcpinto), but if we need to solve this, we could grab the post type from the ID.

```php
get_post_type( $queries['post'] );
```

octaedro replied:
Nice catch @joshuatf! Thank you 🙂
I added the method you mentioned in the commit 2636420

joshuatf reviewed (commented):
Awesome- nice refactor, @octaedro!  This is testing pretty well for me; left one comment about individual posts and a possible fix if we need to include these for tracking purposes.

octaedro committed:
Created method `get_screen_name`

This commit moves the code related  to getting the screen name to a specific method.

octaedro committed:
Added screen name recording for individual posts

This commits adds screen name recording for individual posts

pmcpinto commented:
> I'm not sure how important this is for tracking (feel free to weigh in @pmcpinto), but if we need to solve this, we could grab the post type from the ID.

I'm ok with skipping the identification of specific products/orders. @octaedro @timmyc can this be included on the 4.3 release?

joshuatf reviewed (approved):
Looks great and is testing well!  I'm sure the screen event will prove very useful.  Nice job, @octaedro 🎉

timmyc commented:
cherry-picked to `release/1.3.0-rc.2`
